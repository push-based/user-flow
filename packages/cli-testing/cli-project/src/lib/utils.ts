import * as fs from 'fs';
import { ProcessParams } from './types';
import * as path from 'path';

export function getFolderContent(folders: string[]): string[] {
  return folders.flatMap((d) => {
    if (fs.existsSync(d)) {
      const files = fs.readdirSync(d);
      return files.map((f) => path.join(d, f));
    }
    return [d];
  });
}

export function deleteFileOrFolder(files: string | string[]): void {
  (typeof files === 'string' ? [files] : files).forEach((p) => {
    if (fs.existsSync(p)) {
      const stats = fs.lstatSync(p);
      if(stats.isDirectory()) {
        fs.rmdirSync(p, { recursive: true });
      } else {
        fs.rmSync(p);
      }
      // console.info(`Deleted file ${file}`);
    } else {
      // console.error(`File ${file} does not exist`);
    }
  });
}

export function processParamsToParamsArray(params: ProcessParams): string[] {
  return Object.entries(params).flatMap(([key, value]) => {
    if (key === '_') {
      return value.toString();
    } else if (Array.isArray(value)) {
      return value.map(v => `--${key}=${v.toString()}`);
    } else {
      if (typeof value === 'string') {
        return [`--${key}=${value + ''}`];
      } else if (typeof value === 'boolean') {
        return [`--${value ? '' : 'no-'}${key}`];
      }
      return [`--${key}=${value + ''}`];
    }
  }) as string[];
}

