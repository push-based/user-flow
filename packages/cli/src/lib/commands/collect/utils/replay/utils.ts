import { CustomStep, Step, StepType } from '@puppeteer/replay';
import { MeasureModes, UserFlowRecordingStep } from './types';

export function isMeasureType(str: string) {
  switch (str as MeasureModes) {
    case 'navigate':
    case 'snapshot':
    case 'startTimespan':
    case 'endTimespan':
      return true;
    default:
      return false;
  }
}

export function stringify(enrichedRecordingJson: {
  title: string;
  steps: UserFlowRecordingStep[];
}): string {
  const { title, steps } = enrichedRecordingJson;
  const standardizedJson = {
    title,
    steps: steps.map((step) => {
      if (isMeasureType(step.type)) {
        return userFlowStepToCustomStep(
          step as unknown as UserFlowRecordingStep
        );
      }
      return step;
    }),
  };
  return JSON.stringify(standardizedJson);
}

function userFlowStepToCustomStep(step: UserFlowRecordingStep): Step {
  const { type: name, parameters } = step as any;
  const stdStp: CustomStep = {
    type: StepType.CustomStep,
    name,
    parameters,
  };
  return stdStp;
}
