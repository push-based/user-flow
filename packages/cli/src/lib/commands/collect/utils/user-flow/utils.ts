import { toFileName } from '../../../../core/file';
