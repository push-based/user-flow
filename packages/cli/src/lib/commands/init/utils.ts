import { InitArgvOptions } from './options/types';
import { CollectRcOptions, PersistRcOptions } from '../collect/options/types';
import { AssertRcOptions } from '../assert/options/types';

export function getInitCommandOptionsFromArgv(argv: any) {
  let {
    generateFlow, generateGhWorkflow,
    url, ufPath, serveCommand, awaitServeStdout,
    outPath, format, budgetPath, budgets
  } = argv as unknown as InitArgvOptions;

  let collect = {} as CollectRcOptions;
  url && (collect.url = url);
  ufPath && (collect.ufPath = ufPath);
  // optional
  serveCommand && (collect.serveCommand = serveCommand);
  awaitServeStdout && (collect.awaitServeStdout = awaitServeStdout);

  let persist = {} as PersistRcOptions;
  outPath && (persist.outPath = outPath);
  format && (persist.format = format);

  let assert = {} as AssertRcOptions;
  budgetPath && (assert.budgetPath = budgetPath);
  budgets && (assert.budgets = budgets);

  return { collect, persist, assert, generateFlow, generateGhWorkflow };
}

