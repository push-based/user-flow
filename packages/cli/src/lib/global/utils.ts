import { GlobalOptionsArgv } from './options/types';

export function getGlobalOptionsFromArgv(
  argv: any
): Partial<GlobalOptionsArgv> {
  const { rcPath, interactive, verbose } = argv;

  let globalOptions = {} as GlobalOptionsArgv;
  rcPath && (globalOptions.rcPath = rcPath);
  interactive && (globalOptions.interactive = interactive);
  verbose && (globalOptions.verbose = verbose);

  return globalOptions;
}
