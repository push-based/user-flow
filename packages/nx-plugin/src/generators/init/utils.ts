import { getWorkspaceLayout, readProjectConfiguration, Tree, updateJson } from '@nrwl/devkit';
import {join} from "path";
import {NormalizedSchema} from "./types";
import {InitGeneratorSchema} from "./schema";
import {PLUGIN_NAME} from "../constants";
import {DEFAULT_TARGET_NAME} from "../target/constants";

export function normalizeOptions(tree: Tree, options?: InitGeneratorSchema): NormalizedSchema {

  const projectName = options.projectName;
  const projectRoot = readProjectConfiguration(tree, options.projectName).root;

  return {
    ...options,
    projectName,
    projectRoot
  };
}

export function updateDependencies(tree: Tree, options?: NormalizedSchema) {
  updateJson(tree, join('package.json'), (json) => {
    if (!json.devDependencies) {
      json.devDependencies = {};
    }
    json.devDependencies['@push-based/user-flow'] = '^0.19.0';
    json.devDependencies[PLUGIN_NAME] = '^0.0.0';
    return json;
  });
}

export function updateNxJson(tree: Tree, options?: NormalizedSchema) {
  updateJson(tree, join('nx.json'), (json) => {
    if (json.tasksRunnerOptions) {
      if (!json.tasksRunnerOptions.default.options.cacheableOperations) {
        json.tasksRunnerOptions.default.options.cacheableOperations = []
      }
      json.tasksRunnerOptions.default.options.cacheableOperations.push(DEFAULT_TARGET_NAME);
    }

    if (json.generators === undefined) {
      json.generators = {};
    }
    json.generators[PLUGIN_NAME + ":target"] = {
      "targetName": DEFAULT_TARGET_NAME
    }

    return json;
  });
}
