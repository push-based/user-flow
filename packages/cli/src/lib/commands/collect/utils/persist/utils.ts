import { ReducedReport } from '../report/types';
import { userFlowReportToMdTable } from '../../../..';
import { toFileName } from '../../../../core/file';

export function generateStdoutReport(flowResult: ReducedReport): string {
  const dateTime = new Date().toISOString().replace('T', ' ').split('.')[0].slice(0, -3);
  const mdTable = userFlowReportToMdTable(flowResult);
  return `# ${flowResult.name}\n\nDate/Time: ${dateTime}\n\n${mdTable}`;
}

export function generateMdReport(flowResult: ReducedReport): string {
  const dateTime = new Date().toISOString().replace('T', ' ').split('.')[0].slice(0, -3);
  const mdTable = userFlowReportToMdTable(flowResult);
  return `# ${flowResult.name}\n\nDate/Time: ${dateTime}\n\n${mdTable}`;
}

export function dateToIsoLikeString(date: Date): string {
  return isoDateStringToIsoLikeString(date.toISOString());
}
export function isoDateStringToIsoLikeString(isoDate: string): string {
  return isoDate.replace(/[\-:]/gm, '').split('.').shift() as string;
}

export function toReportName(url: string, flowName: string, date?: string): string {
  date = date || new Date().toISOString();
  return `${toFileName(url)}-${toFileName(flowName)}-${isoDateStringToIsoLikeString(date)}`;
}
