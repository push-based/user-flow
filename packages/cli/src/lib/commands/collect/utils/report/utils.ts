import FlowResult from 'lighthouse/types/lhr/flow';
import { ReducedReport } from '../report/types';
import { parseSteps } from './lh-utils';
import { toFileName } from '../../../../core/file';
import { isoDateStringToIsoLikeString } from '../persist/utils';

export function createReducedReport(flowResult: FlowResult): ReducedReport {
  const steps = parseSteps(flowResult.steps);
  return {
    name: flowResult.name,
    fetchTime: steps[0].fetchTime,
    steps,
  };
}

export function enrichReducedReportWithBaseline(
  reducedReport: ReducedReport,
  baselineReport: FlowResult
): ReducedReport {
  const baselineReducedReport = createReducedReport(baselineReport);
  const baselineResults = Object.fromEntries(
    baselineReducedReport.steps.map((step) => [step.name, step.results])
  );
  const steps = reducedReport.steps.map((step) => ({
    ...step,
    baseline: baselineResults[step.name],
  }));
  return {
    ...reducedReport,
    steps,
  };
}

export function toReportName(
  url: string,
  flowName: string,
  report: ReducedReport
): string {
  const fetchTime = isoDateStringToIsoLikeString(report.steps[0].fetchTime);
  return `${toFileName(url)}-${toFileName(
    flowName
  )}-${isoDateStringToIsoLikeString(fetchTime)}`;
}
