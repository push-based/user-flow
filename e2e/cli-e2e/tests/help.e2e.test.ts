import { describe, expect, it } from 'vitest';
import { executeProcess } from '@code-pushup/utils';

describe('user-flow help', () => {
  it('should print global help', async () => {

    const { code, stdout, stderr } = await executeProcess({
      command: 'user-flow',
      args: ['help']
    });

    expect(code).toBe(0);
    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();
  })
})
