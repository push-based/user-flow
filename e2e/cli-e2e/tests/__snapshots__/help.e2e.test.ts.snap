// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`user-flow help > should print global help 1`] = `
"cli.js

Run a set of user flows and save the result

Commands:
  cli.js init     Setup .user-flowrc.json
  cli.js collect  Run a set of user flows and save the result
  cli.js          Run a set of user flows and save the result          [default]
  cli.js assert   Setup .user-flowrc.json

Options:
      --version           Show version number                          [boolean]
  -v, --verbose           Run with verbose logging    [boolean] [default: false]
  -p, --rcPath            Path to user-flow.config.json. e.g.
                          \`./user-flowrc.json\`
                                         [string] [default: ".user-flowrc.json"]
  -i, --interactive       When false questions are skipped with the values from
                          the suggestions. This is useful for CI integrations.
                                                       [boolean] [default: true]
  -t, --url               URL to analyze                                [string]
  -u, --ufPath            folder containing user-flow files to run. (\`*.uf.ts\`
                          or \`*.uf.js\`)                                 [string]
  -l, --config            Lighthouse configuration (RC file only)       [string]
  -c, --configPath        Path to Lighthouse configuration e.g config.json
                                                                        [string]
  -s, --serveCommand      The npm command to serve your application. e.g. "npm
                          run serve:prod"                               [string]
  -w, --awaitServeStdout  A string in stdout resulting from serving the app, to
                          be awaited before start running the tests. e.g.
                          "server running..."                           [string]
  -d, --dryRun            Execute commands without effects             [boolean]
  -o, --outPath           output folder for the user-flow reports       [string]
  -f, --format            Report output formats e.g. JSON
                               [array] [choices: "html", "json", "md", "stdout"]
  -e, --openReport        Opens browser automatically after the user-flow is
                          collected. (true by default) [boolean] [default: true]
  -k, --budgetPath        Path to budgets.json                          [string]
  -j, --budgets           Performance budgets (RC file only)             [array]
  -h, --help              Show help                                    [boolean]

Examples:
  init  Setup user-flows over prompts
"
`;
